File: app/src/androidTestGoogleplay/java/com/todoroo/astrid/gtasks/GtasksMetadataServiceTest.java
/**
 * Copyright (c) 2012 Todoroo Inc
 *
 * <p>See the file "LICENSE" for the full license governing this code.
 */
package com.todoroo.astrid.gtasks;

import static junit.framework.Assert.assertNotNull;
import static junit.framework.Assert.assertNull;

import android.support.test.runner.AndroidJUnit4;
import com.todoroo.astrid.dao.TaskDao;
import com.todoroo.astrid.data.Task;
import javax.inject.Inject;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.tasks.data.GoogleTask;
import org.tasks.data.GoogleTaskDao;
import org.tasks.injection.InjectingTestCase;
import org.tasks.injection.TestComponent;

@SuppressWarnings("nls")
@RunWith(AndroidJUnit4.class)
public class GtasksMetadataServiceTest extends InjectingTestCase {

  @Inject TaskDao taskDao;
  @Inject GoogleTaskDao googleTaskDao;
  private Task task;
  private GoogleTask metadata;

  @Override
  protected void inject(TestComponent component) {
    component.inject(this);
  }

  @Test
  public void testMetadataFound() {
    givenTask(taskWithMetadata(null));

    whenSearchForMetadata();

    thenExpectMetadataFound();
  }

  @Test
  public void testMetadataDoesntExist() {
    givenTask(taskWithoutMetadata());

    whenSearchForMetadata();

    thenExpectNoMetadataFound();
  }

  private void thenExpectNoMetadataFound() {
    assertNull(metadata);
  }

  private void thenExpectMetadataFound() {
    assertNotNull(metadata);
  }

  // --- helpers

  private void whenSearchForMetadata() {
    metadata = googleTaskDao.getByTaskId(task.getId());
  }

  private Task taskWithMetadata(String id) {
    Task task = new Task();
    task.setTitle("cats");
    taskDao.createNew(task);
    GoogleTask metadata = new GoogleTask(task.getId(), "");
    if (id != null) {
      metadata.setRemoteId(id);
    }
    metadata.setTask(task.getId());
    googleTaskDao.insert(metadata);
    return task;
  }

  private void givenTask(Task taskToTest) {
    task = taskToTest;
  }

  private Task taskWithoutMetadata() {
    Task task = new Task();
    task.setTitle("dogs");
    taskDao.createNew(task);
    return task;
  }
}

File: app/src/androidTest/java/com/todoroo/astrid/gtasks/GtasksMetadataServiceTest.kt
/*
 * Copyright (c) 2012 Todoroo Inc
 *
 * See the file "LICENSE" for the full license governing this code.
 */
package com.todoroo.astrid.gtasks

import com.todoroo.astrid.dao.TaskDao
import com.todoroo.astrid.data.Task
import dagger.hilt.android.testing.HiltAndroidTest
import dagger.hilt.android.testing.UninstallModules
import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Test
import org.tasks.data.GoogleTaskDao
import org.tasks.injection.InjectingTestCase
import org.tasks.injection.ProductionModule
import javax.inject.Inject

@UninstallModules(ProductionModule::class)
@HiltAndroidTest
class GtasksMetadataServiceTest : InjectingTestCase() {
    @Inject lateinit var taskDao: TaskDao
    @Inject lateinit var googleTaskDao: GoogleTaskDao
    
    private var task: Task? = null
    private var metadata: GoogleTask? = null

    @Test
    fun testMetadataFound() = runBlocking {
        givenTask(taskWithMetadata(null))
        whenSearchForMetadata()
        thenExpectMetadataFound()
    }

    @Test
    fun testMetadataDoesntExist() = runBlocking {
        givenTask(taskWithoutMetadata())
        whenSearchForMetadata()
        thenExpectNoMetadataFound()
    }

    private fun thenExpectNoMetadataFound() {
        assertNull(metadata)
    }

    private fun thenExpectMetadataFound() {
        assertNotNull(metadata)
    }

    // --- helpers
    private suspend fun whenSearchForMetadata() {
        metadata = googleTaskDao.getByTaskId(task!!.id)
    }

    private suspend fun taskWithMetadata(id: String?): Task {
        val task = Task()
        task.title = "cats"
        taskDao.createNew(task)
        val metadata = GoogleTask(task.id, "")
        if (id != null) {
            metadata.remoteId = id
        }
        metadata.task = task.id
        googleTaskDao.insert(metadata)
        return task
    }

    private fun givenTask(taskToTest: Task) {
        task = taskToTest
    }

    private suspend fun taskWithoutMetadata(): Task {
        val task = Task()
        task.title = "dogs"
        taskDao.createNew(task)
        return task
    }
}